routemaster/state_machine/api.py

```diff
@@ -113,186 +89,147 @@ def create_label(app: App, label: Label, metadata: Metadata) -> Metadata:
+    could_progress = True
+
+    while could_progress:
+        with app.db.begin() as conn:
+            lock_label(label, conn)
+            current_state = get_current_state(label, state_machine, conn)
+
+            if isinstance(current_state, Action):
+                could_progress = process_action(
+                    app,
+                    current_state,
+                    label,
+                    conn,
+                )
+
+            elif isinstance(current_state, Gate):  # pragma: no branch
+                if not current_state.trigger_on_entry:
+                    return
+
+                could_progress = process_gate(
+                    app,
+                    current_state,
+                    label,
+                    conn,
+                )
+
+            else:
+                raise RuntimeError(  # pragma: no cover
+                    "Unsupported state type {0}".format(current_state),
+                )
+
+
+def delete_label(app: App, label: LabelRef) -> None:
     """
     Deletes the metadata for a label and marks the label as deleted.
 
     The history for the label is not changed (in order to allow post-hoc
     analysis of the path the label took through the state machine).
     """
-    state_machine = _get_state_machine(app, label)
+    state_machine = get_state_machine(app, label)
 
-    metadata_field = labels.c.metadata
     label_filter = and_(
         labels.c.name == label.name,
         labels.c.state_machine == state_machine.name,
     )
 
     with app.db.begin() as conn:
         existing_metadata = conn.scalar(
-            select([metadata_field]).where(label_filter),
+            select([labels.c.metadata]).where(label_filter),
         )
         if existing_metadata is None:
             return
 
+        # Record the label as having been deleted and remove its metadata
         conn.execute(labels.update().where(label_filter).values(
             metadata={},
             deleted=True,
         ))
 
-        _exit_state_machine(label, conn)
-
-
-def _exit_state_machine(
-    label: Label,
-    conn,
-) -> None:
-    current_state_name = conn.scalar(
-        select([history.c.new_state]).where(and_(
-            history.c.label_name == label.name,
-            history.c.label_state_machine == label.state_machine,
-        )).order_by(
-            history.c.created.desc(),
-        ).limit(1)
-    )
-
-    conn.execute(history.insert().values(
-        label_name=label.name,
-        label_state_machine=label.state_machine,
-        old_state=current_state_name,
-        new_state=None,
-    ))
-
-
-def _get_state_machine(app: App, label: Label) -> StateMachine:
-    try:
-        return app.config.state_machines[label.state_machine]
-    except KeyError as k:
-        raise UnknownStateMachine(label.state_machine)
+        # Add a history entry for the deletion
+        current_state_name = conn.scalar(
+            select([history.c.new_state]).where(and_(
```

Comment:
Re-use `label_filter` here?

Reply:
In fact, doesn't this duplicate the logic from `get_current_state`?

Reply:
4fd4688e1c9584f5f995725f50e11cdc934e6166

e6c2d7033273d76e2d649264333eb4208f513a80